### ISSUE TITLE
apt module does not update cache before checking release exists

### ISSUE THREAD

**Issue by ghost** [ansible/ansible#28991]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
`apt` module

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

Debian 8 (`jessie`)

##### SUMMARY

The `apt` module, when asked to ensure a package is installed (`present`) fails if the release (specified via `default_release`) 

##### STEPS TO REPRODUCE

Use the following tasks in a playbook/role with Ansible 2.2.1.0, where at least the host being managed is Debian 8.

```yaml
- name: Add jessie-backports APT repository
  apt_repository:
    filename: "jessie-backports"
    repo: "deb http://httpredir.debian.org/debian jessie-backports main"
    state: present
    update_cache: yes

- name: Install unscd
  apt:
    name: "unscd"
    default_release: "jessie-backports"
    state: present
    update_cache: yes
    cache_valid_time: "900"
```
NB: Presumably other Debian and Ubuntu versions would also be effected, but the example package and/or release may need changing.

##### EXPECTED RESULTS

On a run where both tasks succeed, the first (`apt_repository`) module should update the APT cache because it has just added the APT repository, and as a result the second (`apt`) task should succeed in installing the `unscd` package from the `jessie-backports` release, which is provided by the APT repository added in the first task.

Meanwhile a scenario could occur where the first (`apt_repository`) task adds the APT repository but fails to update the APT cache due to a temporary problem like a lack of Internet connectivity or the repository being down. On a subsequent run (after 15 mins / `900` seconds) after the temporary problem is resolved, the first task will understandably not update APT cache as the APT repository is already added. However the second task should still update the APT cache as it's over 900 seconds old, after which it should be able to install `unscd` from the `jessie-backports` release provided by that APT repository.

##### ACTUAL RESULTS

In the second scenario above, it appears the `apt` module does not update the APT cache before checking the `jessie-backports` and so fails with the below error.

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: SystemError: E:The value 'jessie-backports' is invalid for APT::Default-Release as such a release is not available in the sources
fatal: [example-host]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_EMdwuK/ansible_module_apt.py\", line 903, in <module>\n    main()\n  File \"/tmp/ansible_EMdwuK/ansible_module_apt.py\", line 806, in main\n    cache.open(progress=None)\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 154, in open\n    self._cache = apt_pkg.Cache(progress)\nSystemError: E:The value 'jessie-backports' is invalid for APT::Default-Release as such a release is not available in the sources\n", "module_stdout": "", "msg": "MODULE FAILURE"} 
```
I can appreciate the `apt` module, at least when told a package should be `present`, wants to check for the presence/absence of the package before updating the APT cache, in order to avoid updating the APT cache if the package is already present. However (at least for `state: present`) I would think it should be possible to check whether the named package is installed locally, without concern as to whether the release specified to `default_release` is available. I *would* expect the `apt` module to check the release being avalaible, but only after it's updated the APT cache (if asked to do so!).

**Comment by ansibot** [comment]
cc @mgwilliams
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @mgwilliams
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
@rob-parker-geo are you able to try out the latest devel version and see if the behaviour is still there?

**Comment by krazik** [comment]
temp fix for problem is to just run apt update via shell

```
  tasks:
    - name: Run apt update via shell # Until https://github.com/ansible/ansible/issues/28991
      shell: apt update
      become: yes
```

**Comment by ansibot** [comment]
@rob-parker-geo This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@rob-parker-geo You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by bdkakdb** [comment]
@jborean93 still there in 2.5.1

**Comment by bdkakdb** [comment]
Looks like the same:

https://github.com/ansible/ansible/issues/30754

**Comment by ph448** [comment]
This is still happening with Ansible 2.7.11.
I disagree with @bdkakdb that it's the same issue. If I run `apt update` manually or use a separate `apt` task to only update the cache before the one installing packages, package installation succeeds, but Ansible does not seem to successfully execute the cache update correctly on tasks that also install packages.

**Comment by danvaida** [comment]
FWIW, the symptom described here is still present on `2.9.15`. Haven't checked yet if the cause is actually the same.

**Comment by s-hertel** [comment]
I'm able to reproduce this on recent versions. `apt_repository` updates the cache by default, so I had to set `update_cache: False` for the setup task in order to reproduce the issue with the `apt` module.

```
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1695232371.9362087-410-134629965927984/AnsiballZ_apt.py", line 107, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1695232371.9362087-410-134629965927984/AnsiballZ_apt.py", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1695232371.9362087-410-134629965927984/AnsiballZ_apt.py", line 47, in invoke_module
    runpy.run_module(mod_name='ansible.modules.apt', init_globals=dict(_module_fqn='ansible.modules.apt', _modlib_path=modlib_path),
  File "/usr/lib/python3.9/runpy.py", line 210, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_apt_payload_ncnerd_v/ansible_apt_payload.zip/ansible/modules/apt.py", line 1512, in <module>
  File "/tmp/ansible_apt_payload_ncnerd_v/ansible_apt_payload.zip/ansible/modules/apt.py", line 1342, in main
  File "/usr/lib/python3/dist-packages/apt/cache.py", line 214, in open
    self._cache = apt_pkg.Cache(progress)
apt_pkg.Error: E:The value 'bullseye-backports' is invalid for APT::Default-Release as such a release is not available in the sources
```

The issue is specifically setting the Default-Release before the cache update https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/apt.py#L1353-L1359

Maybe we need to update the cache twice - first with the existing config, and then with the default release configured if one is specified.
